clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2538,13 +2538,16 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
     return CompoundStmt::Create(SemaRef.getASTContext(), BodyStmts, {}, {});
   }
 
-  void markParallelWorkItemCalls() {
+  void markParallelWorkItemCalls(SourceLocation SrcLoc) {
     if (getKernelInvocationKind(KernelCallerFunc) ==
         InvokeParallelForWorkGroup) {
       FindPFWGLambdaFnVisitor V(KernelObj);
       V.TraverseStmt(KernelCallerFunc->getBody());
       CXXMethodDecl *WGLambdaFn = V.getLambdaFn();
-      assert(WGLambdaFn && "PFWG lambda not found");
+      if (!WGLambdaFn) {
```

Comment:
So, did I get it right, that this change issues an error when callable object passed to `parallel_for_work_group` does not satisfy the requirements implemented by `FindPFWGLambdaFnVisitor::VisitCallExpr`?
Looking at them, they are mostly checking that call operator of the corresponding class receive `sycl::group` parameter and may or may not receive an optional `kernel_handler` parameter. I wonder why this is not enforced by the code in headers instead?
I mean wouldn't the following code:
```
  template <typename KernelName, typename ElementType, typename KernelType>
__attribute__((sycl_kernel))
  kernel_parallel_for_work_group(KernelType KernelFunc) {
  sycl::group G; // Somehow declare the group, it doesn't matter how
  KernelFunc(G);
}
```
make FE reject any callable object as `KernelFunc` argument if its call operator doesn't accept `sycl::group` parameter?
Thinking about that, I'm not sure why we need an additional error emission.

Reply:
In fact, from the very brief look, I don't see why we need `FindPFWGLambdaFnVisitor` at all. Why the kernel object cannot be taken from kernel caller function parameters via `GetSYCLKernelObjectType`?

Reply:
> So, did I get it right, that this change issues an error when callable object passed to `parallel_for_work_group` does not satisfy the requirements implemented by `FindPFWGLambdaFnVisitor::VisitCallExpr`? Looking at them, they are mostly checking that call operator of the corresponding class receive `sycl::group` parameter and may or may not receive an optional `kernel_handler` parameter. I wonder why this is not enforced by the code in headers instead? I mean wouldn't the following code:
> 
> ```
>   template <typename KernelName, typename ElementType, typename KernelType>
> __attribute__((sycl_kernel))
>   kernel_parallel_for_work_group(KernelType KernelFunc) {
>   sycl::group G; // Somehow declare the group, it doesn't matter how
>   KernelFunc(G);
> }
> ```
> 
> make FE reject any callable object as `KernelFunc` argument if its call operator doesn't accept `sycl::group` parameter? Thinking about that, I'm not sure why we need an additional error emission.

The signatures for parallel_for_work_group is defined in the spec, and I am not sure if this can be changed to avoid this error.  The group is under user-control from what I understand; not quite sure what interface you are proposing to address this.

> I don't see why we need FindPFWGLambdaFnVisitor at all.
>  Why the kernel object cannot be taken from kernel caller function parameters via GetSYCLKernelObjectType?

The difference seems to be that FPWG visitor operates on the call site (by walking the caller's function body) and fetches the lambda function passed there.   GetSYCLKernelObjectType, OTOH, seems to check the function parameters for type matching.

This seems to have been implemented long back: https://github.com/intel/llvm/pull/345 and refactored here: https://github.com/intel/llvm/pull/1517 (by you! :-)), so I am a bit wary of redesigning this on the fly.

But unless the PFWG interface is completely removed, we will still get to a situation where we need the Lambda function and I am only addressing the case where we expect it to be found, but are not for some reason.   Currently, we are asserting and the crash looks bad for the compiler and does not really help the user.   Hence the diagnostic.

Reply:
> This seems to have been implemented long back: https://github.com/intel/llvm/pull/345 and refactored here: https://github.com/intel/llvm/pull/1517 (by you! :-)), so I am a bit wary of redesigning this on the fly.

I think it is yet another bad design choice introduced while ago. I would prefer refactoring bad code instead of fixing it. https://github.com/intel/llvm/pull/1517 was quite big refactoring already, so it didn't change this part at all, just moved to another place.

> The signatures for parallel_for_work_group is defined in the spec, and I am not sure if this can be changed to avoid this error. The group is under user-control from what I understand; not quite sure what interface you are proposing to address this.

My point was that the only difference between FPWG visitor and `GetSYCLKernelObjectType` is that the visitor doesn't consider callable objects as valid if they don't receive group parameter. If SYCL headers declare `kernel_parallel_for_work_group` function the way I mentioned (or similar), this function will reject any passed callable object whose `()` operator doesn't receive `sycl::group`. And they will likely do that, because the spec says that the parameter of callable object is `sycl::group` object. So it is ok to not check on that in SemaSYCL.cpp and access the lambda via ``GetSYCLKernelObjectType`
I was able to perform an experiment locally - I removed the visitor and changed it with:
```
-      FindPFWGLambdaFnVisitor V(KernelObj);
-      V.TraverseStmt(KernelCallerFunc->getBody());
-      CXXMethodDecl *WGLambdaFn = V.getLambdaFn();
-      assert(WGLambdaFn && "PFWG lambda not found");
+      CXXRecordDecl *KernelObj=
+          GetSYCLKernelObjectType(KernelCallerFunc)->getAsCXXRecordDecl();
+      CXXMethodDecl *WGLambdaFn;
+      if (KernelObj->isLambda())
+        WGLambdaFn = KernelObj->getLambdaCallOperator();
+      else
+        WGLambdaFn = getOperatorParens(KernelObj);
+      assert(WGLambdaFn && "non callable object is passed as kernel obj");
```
Not only that removes the crash, but also passes simple tests on hierarchical parallelism I took from inte/llvm-test-suite repo.

Reply:
> the spec says that the parameter of callable object is sycl::group object. So it is ok to not check on that in SemaSYCL.cpp

Thanks, this gives me better confidence for making the change you are proposing.
Thank you very much for the experimentation too!

Reply:
Thank you for the change! I'm really glad this worked out.